components/brave_ads/content/browser/search_result_ad/search_result_ad_redirect_throttle.cc

```diff
@@ -0,0 +1,123 @@
+#include "url/gurl.h"
+
+namespace brave_ads {
+
+namespace {
+
+SessionID GetTabId(content::WebContents* web_contents) {
+  DCHECK(web_contents);
+
+  SessionID tab_id = sessions::SessionTabHelper::IdForTab(web_contents);
+  content::RenderFrameHost* opener_rfh = web_contents->GetOpener();
+  if (!opener_rfh) {
+    opener_rfh = web_contents->GetOriginalOpener();
+  }
+  if (opener_rfh) {
+    content::WebContents* original_web_contents =
+        content::WebContents::FromRenderFrameHost(opener_rfh);
+    tab_id = sessions::SessionTabHelper::IdForTab(original_web_contents);
+  }
+
+  return tab_id;
+}
+
+}  // namespace
+
+// static
+std::unique_ptr<SearchResultAdRedirectThrottle>
+SearchResultAdRedirectThrottle::MaybeCreateThrottleFor(
+    SearchResultAdService* search_result_ad_service,
+    const network::ResourceRequest& request,
+    content::WebContents* web_contents) {
+  DCHECK(web_contents);
+
+  if (!search_result_ad_service || !request.request_initiator ||
+      !request.has_user_gesture) {
+    return nullptr;
+  }
+
+  // !request.is_main_frame
+  if (request.resource_type !=
+      static_cast<int>(blink::mojom::ResourceType::kMainFrame)) {
+    return nullptr;
+  }
+
+  if (!brave_search::IsAllowedHost(request.request_initiator->GetURL())) {
+    return nullptr;
+  }
+
+  if (!IsSearchResultAdClickedConfirmationUrl(request.url)) {
+    return nullptr;
+  }
+
+  SessionID tab_id = GetTabId(web_contents);
+  if (!tab_id.is_valid()) {
+    return nullptr;
+  }
+
+  return std::make_unique<SearchResultAdRedirectThrottle>(
+      search_result_ad_service, tab_id);
+}
+
+SearchResultAdRedirectThrottle::SearchResultAdRedirectThrottle(
+    SearchResultAdService* search_result_ad_service,
+    SessionID tab_id)
+    : search_result_ad_service_(search_result_ad_service), tab_id_(tab_id) {
+  DCHECK(search_result_ad_service_);
+  DCHECK(tab_id_.is_valid());
+}
+
+SearchResultAdRedirectThrottle::~SearchResultAdRedirectThrottle() = default;
+
+void SearchResultAdRedirectThrottle::WillStartRequest(
+    network::ResourceRequest* request,
+    bool* defer) {
+  DCHECK(request);
+  DCHECK(request->request_initiator);
+  DCHECK(brave_search::IsAllowedHost(request->request_initiator->GetURL()));
+
+  std::string creative_instance_id =
+      GetClickedSearchResultAdCreativeInstanceId(request->url);
+  if (creative_instance_id.empty()) {
+    return;
+  }
+
+  absl::optional<GURL> search_result_ad_target_url =
+      search_result_ad_service_->MaybeTriggerSearchResultAdClickedEvent(
+          creative_instance_id, tab_id_);
+  if (!search_result_ad_target_url) {
+    return;
+  }
+
+  url::Origin original_origin = url::Origin::Create(request->url);
+  url::Origin search_result_ad_target_origin =
+      url::Origin::Create(*search_result_ad_target_url);
+  request->url = *search_result_ad_target_url;
+
+  if (!original_origin.IsSameOriginWith(search_result_ad_target_origin)) {
```

Comment:
what is the purpose of this?

Reply:
This code checks if new url origin differs from old one. If so, then it updates network request parameters to fit new origin.

Reply:
What we do here is making redirect from `original_origin` to the `search_result_ad_target_origin`. 
Initially `site_for_cookies` represents `original_origin` as same-site for the network request. As we redirect, we need to change `site_for_cookies` to represent new same-site origin `search_result_ad_target_origin`.
So we are navigating to `search_result_ad_target_origin` and reading cookies for `search_result_ad_target_origin`.
Another variant is to make `site_for_cookies` empty to don't match any origins at all.

Reply:
thanks for the explanation. the empty option seems better to me but i'll defer to @bridiver

Reply:
We absolutely should not be messing with site for cookies. It exists specifically to prevent doing this and we should not be making security exceptions for ourselves like this.

Reply:
what exactly are you trying to accomplish @aseren ? And I'm not asking about site for cookies, I'm asking what the end goal is.

Reply:
Discussed in DM. Removed this code.